### PR TITLE
Load symfony container in front

### LINKED ISSUE #31846: Spike: Make Symfony Container accessible in Front Office
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

So far the Symfony container is only accessible in the Back Office, as such many services from Symfony are not usable in Front Office Even though we don't plan on starting the migration of the front in Symfony right now we could still anticipate at least the access of the container so that some services can be used in front as well among which:
- the CQRS command bus
- the future CartRule/Price new computer
- doctrine entities / services (even if today an alternative exists)
- ...

### Expected behavior

- [ ] We need to instanciate the Symfony kernel in the FO endpoint index.php file so it's accessible
- [ ] In the FO we need a clean way to get the container (for now we use SymfonyContainer::getInstance in legacy code in the BO but it should be improvable)
- [ ] We need to test that any service from the BO is also usable in FO
- [ ] We need to perform a benchmark and analyze the cost of this container in FO
  - [ ] When no Symfony services is used
  - [ ] When a simple Symfony service is used
  - [ ] When multiple services are used
- [ ] This feature should be handled via a feature flag so it can be completely disabled

### Steps to reproduce

See previous content

### PrestaShop version(s) where the bug happened

9.0

### PHP version(s) where the bug happened

_No response_

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by boherm**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add symfony container for front and dotenv files for configurations!<br>We can enable this **experimental** feature with new dotenv file!
| Type?             | new feature
| Category?         | FO
| BC breaks?        | yes
| Deprecations?     | no
| How to test?      | Set `PS_CONTAINER_V2_FRONT=true` in `.env` file.
| Fixed ticket?     | Fixes #31846
| Related PRs       | 
| Sponsor company   | 

**Notes: You can run tests ui with this PR: https://github.com/PrestaShop/ga.tests.ui.pr/pull/34.**

**Tests:**
- CI ✅
- [Autotests (without new container activated)](https://github.com/aniszr/ga.tests.ui.pr/actions/runs/6010082838) ⏳
- [Autotests (with new container activated)](https://github.com/boherm/ga.tests.ui.pr/actions/runs/6010749136) ⏳

**Some performances tests:**
*(with PHP 8.1, no opcache, no docker, with apache & mysql 8.0.33 on MacBook Pro M2 16go Ram)*
<table>
    <thead>
        <tr>
            <th>Page</th>
            <th>Dev</th>
            <th>V2</th>
            <th>Time</th>
            <th>I/O wait</th>
            <th>CPU</th>
            <th>Memory</th>
            <th>Network</th>
        </tr>
    </thead>
        <tr>
            <td rowspan="4">FO - Homepage</td>
            <td>❌</td><td>❌</td><td>152 ms</td><td>46.9 ms</td><td>105 ms</td><td>6.3 MB</td><td>205 kB</td></tr>
        <tr><td>❌</td><td>✅</td><td>167 ms</td><td>44.9 ms</td><td>122 ms</td><td>7.5 MB</td><td>210 kB</td></tr>
        <tr><td>✅</td><td>❌</td><td>148 ms</td><td>30.1 ms</td><td>118 ms</td><td>6.3 MB</td><td>205 kB</td></tr>
        <tr><td>✅</td><td>✅</td><td>498 ms</td><td>58.2 ms</td><td>440 ms</td><td>12.1 MB</td><td>210 kB</td></tr>
        <tr>
            <td rowspan="4">FO - Product page</td>
            <td>❌</td><td>❌</td><td>224 ms</td><td>76.2 ms</td><td>148 ms</td><td>8.4 MB</td><td>347 kB</td></tr>
        <tr><td>❌</td><td>✅</td><td>247 ms</td><td>79.5 ms</td><td>167 ms</td><td>9.5 MB</td><td>352 kB</td></tr>
        <tr><td>✅</td><td>❌</td><td>254 ms</td><td>91.5 ms</td><td>162 ms</td><td>8.4 MB</td><td>347 kB</td></tr>
        <tr><td>✅</td><td>✅</td><td>616 ms</td><td>98.4 ms</td><td>518 ms</td><td>14.7 MB</td><td>352 kB</td></tr>
        <tr>
            <td rowspan="4">FO - Products listing</td>
            <td>❌</td><td>❌</td><td>431 ms</td><td>160 ms</td><td>271 ms</td><td>10.4 MB</td><td>650 kB</td></tr>
        <tr><td>❌</td><td>✅</td><td>426 ms</td><td>162 ms</td><td>264 ms</td><td>11.5 MB</td><td>655 kB</td></tr>
        <tr><td>✅</td><td>❌</td><td>418 ms</td><td>160 ms</td><td>258 ms</td><td>10.4 MB</td><td>650 kB</td></tr>
        <tr><td>✅</td><td>✅</td><td>801 ms</td><td>176 ms</td><td>625 ms</td><td>18 MB</td><td>655 kB</td></tr>
        <tr>
            <td rowspan="4">FO - Cart<br/><small><em>when connected with one product</em></small></td>
            <td>❌</td><td>❌</td><td>220 ms</td><td>73.8 ms</td><td>146 ms</td><td>7.5 MB</td><td>317 kB</td></tr>
        <tr><td>❌</td><td>✅</td><td>318 ms</td><td>109 ms</td><td>208 ms</td><td>14.4 MB</td><td>339 kB</td></tr>
        <tr><td>✅</td><td>❌</td><td>207 ms</td><td>64.1 ms</td><td>143 ms</td><td>7.5 MB</td><td>317 kB</td></tr>
        <tr><td>✅</td><td>✅</td><td>638 ms</td><td>88.2 ms</td><td>550 ms</td><td>14.4 MB</td><td>322 kB</td></tr>
        <tr>
            <td rowspan="4">FO - Cart<br/><small><em>when connected with 19 products</em></small></td>
            <td>❌</td><td>❌</td><td>852 ms</td><td>312 ms</td><td>541 ms</td><td>13.8 MB</td><td>1270 kB</td></tr>
        <tr><td>❌</td><td>✅</td><td>884 ms</td><td>330 ms</td><td>554 ms</td><td>15.2 MB</td><td>1270 kB</td></tr>
        <tr><td>✅</td><td>❌</td><td>847 ms</td><td>313 ms</td><td>533 ms</td><td>14.0 MB</td><td>1270 kB</td></tr>
        <tr><td>✅</td><td>✅</td><td>1370 ms</td><td>330 ms</td><td>1040 ms</td><td>28.5 MB</td><td>1270 kB</td></tr>
        <tr>
            <td rowspan="4">FO - Cart<br/><small><em>when connected with 40 products</em></small></td>
            <td>❌</td><td>❌</td><td>1740 ms</td><td>610 ms</td><td>1130 ms</td><td>21,6 MB</td><td>2360 kB</td></tr>
        <tr><td>❌</td><td>✅</td><td>1780 ms</td><td>627 ms</td><td>1160 ms</td><td>22.2 MB</td><td>2360 kB</td></tr>
        <tr><td>✅</td><td>❌</td><td>1690 ms</td><td>588 ms</td><td>1100 ms</td><td>20.9 MB</td><td>2360 kB</td></tr>
        <tr><td>✅</td><td>✅</td><td>2480 ms</td><td>658 ms</td><td>1820 ms</td><td>45.9 MB</td><td>2360 kB</td></tr>
</table>

**Conclusions about performances tests:**
- This new container can slow down some pages.
    - Front migration can optimize performance in some way.
- Use one or multiples services within the container v2, don't get slower the front.
- Use new container with dev mode may slow down PrestaShop in front.

### CODE REVIEW

**Comment by nicosomb** [comment]
FYI, @FabienPapet opened (and closed) this PR https://github.com/PrestaShop/PrestaShop/pull/29997

File: classes/Hook.php [hunk 1]
```diff
@@ -1007,6 +1007,7 @@ public static function coreRenderWidget($module, $hook_name, $params)
         try {
             return $module->renderWidget($hook_name, $params);
         } catch (Exception $e) {
+            dd($hook_name, $params, $e);
```
Comment (FabienPapet) [inline classes/Hook.php]: 👀

Comment (boherm) [reply classes/Hook.php]: Oupsy!

File: src/Adapter/Presenter/Product/ProductLazyArray.php [hunk 2]
```diff
@@ -647,7 +647,7 @@ private function shouldShowOutOfStockLabel(ProductPresentationSettings $settings
         if ($product['id_product_attribute']) {
             // Displayed only if all combinations are out of stock (stock is <= 0)
             $product = new Product((int) $product['id_product']);
-            if (empty($product->id)) {
+            if (empty($product->id) || $product->getAttributesResume($this->language->getId()) === false) {
```
Comment (kpodemski) [inline src/Adapter/Presenter/Product/ProductLazyArray.php]: is it required for this PR? 🤔

File: .env.dist
Comment (lartist) [inline .env.dist:1]: We should remove this file since we already have .env

File: app/AppKernel.php [hunk 3]
```diff
@@ -32,14 +32,28 @@
...
+
+    /**
+     * @var string
+     */
+    protected $appType;
+
+    public function __construct(string $environment, bool $debug)
+    {
+        parent::__construct($environment, $debug);
+    }
```
Comment (FabienPapet) [inline app/AppKernel.php]: useless

File: src/PrestaShopBundle/PrestaShopBundle.php [hunk 4]
```diff
@@ -72,6 +85,8 @@ public function build(ContainerBuilder $container)
     {
         $container->addCompilerPass(new DynamicRolePass());
         $container->addCompilerPass(new PopulateTranslationProvidersPass());
+        $container->addCompilerPass(new LoadServicesFromModulesPass(), PassConfig::TYPE_BEFORE_OPTIMIZATION, self::LOAD_MODULE_SERVICES_PASS_PRIORITY);
+        $container->addCompilerPass(new LoadServicesFromModulesPass($this->appType), PassConfig::TYPE_BEFORE_OPTIMIZATION, self::LOAD_MODULE_SERVICES_PASS_PRIORITY);
```
Comment (FabienPapet) [inline src/PrestaShopBundle/PrestaShopBundle.php]: This is super dangerous. The DI should never know if we are in a frontend or a backend context. This information should come from the request for example, but should not be used inside a compiler pass.

File: app/AdminKernel.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+require_once 'AppKernel.php';
+
+class AdminKernel extends AppKernel
+{
+    public function __construct(string $environment, bool $debug)
+    {
+        parent::__construct($environment, $debug);
+        $this->appType = AppKernel::TYPE_ADMIN;
```
Comment (lartist) [inline app/AdminKernel.php]: I think appType it's not handy. We should remove it and based the comparison on $kernel::class (which should be either AdminKernel or FrontKernel)

File: app/FrontKernel.php
Comment (lartist) [inline app/FrontKernel.php:1]: Where is this class used?

**Comment by Hlavtox** [comment]
FO cart is already extremely to slow, to a point where it's on the edge of usability, we can't offer to have this slowdown :(

Snímek obrazovky 2023-07-18 070844

**Comment by boherm** [comment]
Hi @Hlavtox, yes and it's a big slow.
But, I really don't why, I can't reproduce this time again... Now, I don't have a big time delta like this.

To know what is going on, I want to reinstall a fresh PrestaShop instance and retests some pages like I did.
I wonder if it's a Symfony kernel cache issue that can be really ready after dozen requests.

**Review by tleon — COMMENTED** [review]
LGTM, tho I don't think I know the topic enought to aprouve this on my own. Would love to have other inputs

File: composer.json [hunk 6]
```diff
@@ -237,6 +237,7 @@
         },
         "classmap": [
             "app/AppKernel.php",
+            "app/AdminKernel.php",
```
Comment (mflasquin) [inline composer.json:240]:
```suggestion
            "app/AdminKernel.php",
            "app/FrontKernel.php",
```

File: src/PrestaShopBundle/Resources/config/services/front_legacy.yml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  prestashop.adapter.module.repository.module_repository:
+    class: 'PrestaShop\PrestaShop\Adapter\Module\Repository\ModuleRepository'
+    arguments:
+      - !php/const _PS_ROOT_DIR_
+      - !php/const _PS_MODULE_DIR_
+
+  prestashop.translation.translator_language_loader:
+    class: PrestaShopBundle\Translation\TranslatorLanguageLoader
+    arguments:
+      - '@prestashop.adapter.module.repository.module_repository'
```
Comment (mflasquin) [inline src/PrestaShopBundle/Resources/config/services/front_legacy.yml:57]: Regarding this ADR, can you use FQCN for services names https://github.com/PrestaShop/ADR/blob/master/0019-FQCN-Autowiring.md ?

**Review by matks — APPROVED** [review]
Implementation is OK for me 👍 

There's a little comment from @mflasquin I have nothing to add

**Comment by Hlavtox** [comment]
@matks It cannot be merged if it slows down cart loading 3 times in production mode

File: src/PrestaShopBundle/Resources/config/services/front_legacy.yml (hunk 7 above)
Comment (boherm) [reply src/PrestaShopBundle/Resources/config/services/front_legacy.yml:57]: This is a configuration for legacy services used in front, and I can't really switch to FQCN without some big changes that can introduce some big breaking changes (and if we delete theses alias).
But, I think that can be done in another PR and why not optimize this system too.
What do you think about it?

**Comment by boherm** [comment]
@Hlavtox, after some new benchmarks with this system, I don't be able to reproduce the loading X3 in production mode.
In fact, I've got in all new tests arround 310ms in cart page with this implementation and in production mode.
(I've update the table in description with new values)

So I think that can be ok to merge seen that this implementation is behind a feature flag.

**Review by M0rgan01 — APPROVED** [review]
Thank you for the performance comparison table :) The changes announce a good thing

**Review by kpodemski — APPROVED** [review]

**Comment by boherm** [comment]
[Autotests (without new container activated)](https://github.com/aniszr/ga.tests.ui.pr/actions/runs/6010082838)
[Autotests (with new container activated)](https://github.com/boherm/ga.tests.ui.pr/actions/runs/6010749136)

**Review by aniszr — APPROVED** [review]
Hello @boherm 

I tested your PR and it LGTM!
The modifications done on the.env file didn't cause any problems on the FO:heavy_check_mark: 

See attached screenshots:
image

image

- Auto Test (without container activated): https://github.com/aniszr/ga.tests.ui.pr/actions/runs/6010082838
- Auto Test (container activated): https://github.com/boherm/ga.tests.ui.pr/actions/runs/6010749136
QA :heavy_check_mark: 

Thanks!

**Comment by mflasquin** [comment]
Big thanks to @boherm for this PR !

**Comment by matks** [comment]
@mflasquin Don't forget the milestone

**Comment by mickaelandrieu** [comment]
This was already a feature (see https://github.com/PrestaShop/PrestaShop/blob/develop/classes/container/LegacyCompilerPass.php), you have to tag a service with the tag "front" to make it available in every legacy controller.

What's the difference with this new container ?

An old exemple of integration => https://github.com/friends-of-presta/masterclass/blob/master/config/services.yml#L53

**Comment by PululuK** [comment]
Hummm good point @mickaelandrieu !

**Pull request merged**